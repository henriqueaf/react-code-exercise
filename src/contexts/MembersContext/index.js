import React, { createContext, useEffect, useState, useContext } from 'react';

import { useMembersReducer } from '../../reducers/MembersReducer';
import { setLoading, setMembers, setFilteredMembers, setCurrentPage, setSelectedSession } from '../../reducers/MembersReducer/actions';
import { getMembersCache } from '../../services/membersCache';

const HOUSE_MINIMUM_SESSION = 102;
const SENATE_MINIMUM_SESSION = 80;

const MembersContext = createContext({});

const MembersProvider = ({ children }) => {
  const [state, dispatch] = useMembersReducer();
  const {
    filteredMembers,
    membersPerPage,
    selectedChamber,
    selectedSession,
  } = state;

  const [minimumSession, setMinimumSession] = useState(SENATE_MINIMUM_SESSION);

  const calculateMinimumSession = () => {
    switch (selectedChamber) {
      case 'senate':
        return SENATE_MINIMUM_SESSION;
      case 'house':
        return HOUSE_MINIMUM_SESSION;
    }
  };

  const providerValue = {
    minimumSession,
    state,
    dispatch,
  };

  useEffect(() => {
    dispatch(setLoading(true));

    const url = `https://api.propublica.org/congress/v1/${selectedSession}/${selectedChamber}/members.json`;

    async function membersCache() {
      try {
        const responseMembers = await getMembersCache(url);
        dispatch(setMembers(responseMembers));
        dispatch(setFilteredMembers(responseMembers));
      } catch (error) {
        console.error(error.message);
        dispatch(setLoading(false));
      } finally {
        dispatch(setLoading(false));
        setMinimumSession(calculateMinimumSession());
      }
    }

    membersCache()
  }, [selectedChamber, selectedSession]);

  useEffect(() => {
    if (selectedChamber === 'house' && selectedSession < HOUSE_MINIMUM_SESSION)
      dispatch(setSelectedSession(HOUSE_MINIMUM_SESSION));
  }, [selectedChamber]);

  useEffect(() => {
    dispatch(setCurrentPage(0));
  }, [membersPerPage, filteredMembers]);

  return (
    <MembersContext.Provider value={providerValue}>
      {children}
    </MembersContext.Provider>
  );
};

const useMembersContext = () => useContext(MembersContext);

export { useMembersContext, MembersProvider, MembersContext };
