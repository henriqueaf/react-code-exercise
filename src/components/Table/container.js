import React, {useContext} from 'react';
import { useMembersContext } from '../../contexts/MembersContext';

import Table from './index';

const Container = () => {
  const {
    state: {
      filteredMembers,
      currentPage,
      membersPerPage
    },
    dispatch
  } = useMembersContext();

  return (
    <Table
      filteredMembers={filteredMembers}
      membersPerPage={membersPerPage}
      currentPage={currentPage}
      dispatch={dispatch}
    />
  )
}

export default Container;
