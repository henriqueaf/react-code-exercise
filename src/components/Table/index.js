import React from 'react';
import { Table } from 'react-bootstrap';

import { useMembersContext } from '../../contexts/MembersContext';
import ListItem from '../ListItem';

export default () => {
  const {
    filteredMembers,
    setSelectedMemberForDetails,
    currentPage,
    membersPerPage
  } = useMembersContext();

  const startIndex = currentPage * membersPerPage;
  const endIndex = startIndex + membersPerPage;

  return (
    <Table striped bordered hover>
      <thead>
        <tr>
          <th>#</th>
          <th>Name</th>
          <th>Twitter</th>
          <th>Facebook</th>
          <th>Youtube</th>
          <th>Party</th>
          <th>District/State</th>
          <th>Next Election</th>
          <th>Details</th>
        </tr>
      </thead>
      <tbody>
        {
          filteredMembers.slice(startIndex, endIndex).map((member, index) => (
            <ListItem member={member} key={index} setSelectedMemberForDetails={setSelectedMemberForDetails} />
          ))
        }
      </tbody>
    </Table>
  );
};
