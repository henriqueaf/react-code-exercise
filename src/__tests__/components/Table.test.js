import {render, fireEvent, screen, cleanup} from '@testing-library/react';
import * as MembersContext from '../../contexts/MembersContext';
import Table from '../../components/Table';

const member = {
  id: 'A00123',
  suffix: 'Sr.',
  first_name: 'Lamar',
  middle_name: 'Adam',
  last_name: 'Frank',
  twitter_account: 'lamartwt',
  facebook_account: 'lamarfc',
  youtube_account: 'lamaryoutube',
  party: 117,
  district: 'LA',
  state: 'LA',
  next_election: 2022
}

const contextValues = {
  filteredMembers: [member],
  setSelectedMemberForDetails: () => {},
  membersPerPage: 1,
  currentPage: 0
};

beforeEach(() => {
  jest
  .spyOn(MembersContext, 'useMembersContext')
  .mockImplementation(() => contextValues);
});

afterEach(() => {
  jest.clearAllMocks();
  cleanup();
});

const mount = () => {
  return render(<Table />);
};

describe('Table', () => {
  it('Matches snapshot', () => {
    const {container} = mount();
    expect(container).toMatchSnapshot();
  });

  it('Renders member data on table columns', () => {
    mount();
    const tableRow = screen.getByRole('row', {name: 'A00123 Lamar Frank lamartwt lamarfc lamaryoutube 117 /LA 2022 Details'});
    expect(tableRow.children[0].textContent).toBe(member.id);
  });
})
