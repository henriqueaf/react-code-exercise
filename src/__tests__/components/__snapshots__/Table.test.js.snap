// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Matches snapshot 1`] = `
<div>
  <table
    class="table table-striped table-bordered table-hover"
  >
    <thead>
      <tr>
        <th>
          #
        </th>
        <th>
          Name
        </th>
        <th>
          Twitter
        </th>
        <th>
          Facebook
        </th>
        <th>
          Youtube
        </th>
        <th>
          Party
        </th>
        <th>
          District/State
        </th>
        <th>
          Next Election
        </th>
        <th>
          Details
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          A00123
        </td>
        <td>
           Lamar Frank
        </td>
        <td>
          <a
            href="https://twitter.com/lamartwt"
            target="_blank"
          >
            lamartwt
          </a>
        </td>
        <td>
          <a
            href="https://www.facebook.com/lamarfc"
            target="_blank"
          >
            lamarfc
          </a>
        </td>
        <td>
          <a
            href="https://www.youtube.com/user/lamaryoutube"
            target="_blank"
          >
            lamaryoutube
          </a>
        </td>
        <td>
          117
        </td>
        <td>
          /LA
        </td>
        <td>
          2022
        </td>
        <td>
          <a
            href="#none"
          >
            Details
          </a>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
